hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ContainerEndpoint.java

```diff
@@ -498,72 +535,138 @@ private List<ContainerBlockMetadata> getBlocks(
+   * Storage Container Manager (SCM), based on the provided filter parameter.
+   * The returned list is paginated based on the provided limit and prevKey
+   * parameters.
+   *
+   * @param limit   The maximum number of container discrepancies to return.
+   * @param prevKey The container ID after which the results are returned.
+   * @param missingIn  The missing filter parameter to specify if it's
+   *                   "OM" or "SCM" missing containers to be returned.
+   */
   @GET
   @Path("/mismatch")
-  public Response getContainerMisMatchInsights() {
+  public Response getContainerMisMatchInsights(
+      @DefaultValue(DEFAULT_FETCH_COUNT)
+      @QueryParam(RECON_QUERY_LIMIT) int limit,
+      @DefaultValue(PREV_CONTAINER_ID_DEFAULT_VALUE)
+      @QueryParam(RECON_QUERY_PREVKEY) long prevKey,
+      @DefaultValue(DEFAULT_FILTER_FOR_MISSING_CONTAINERS)
+      @QueryParam(RECON_QUERY_FILTER) String missingIn) {
+    if (prevKey < 0 || limit < 0) {
+      // Send back an empty response
+      return Response.status(Response.Status.NOT_ACCEPTABLE).build();
+    }
+
     List<ContainerDiscrepancyInfo> containerDiscrepancyInfoList =
         new ArrayList<>();
     try {
       Map<Long, ContainerMetadata> omContainers =
           reconContainerMetadataManager.getContainers(-1, -1);
       List<Long> scmNonDeletedContainers =
           containerManager.getContainers().stream()
-              .filter(containerInfo -> !(containerInfo.getState() ==
-                  HddsProtos.LifeCycleState.DELETED))
-              .map(containerInfo -> containerInfo.getContainerID()).collect(
-                  Collectors.toList());
-
-      // Filter list of container Ids which are present in OM but not in SCM.
-      List<Map.Entry<Long, ContainerMetadata>> notSCMContainers =
-          omContainers.entrySet().stream().filter(containerMetadataEntry ->
-                  !(scmNonDeletedContainers.contains(
-                      containerMetadataEntry.getKey())))
-              .collect(
-                  Collectors.toList());
+              .filter(containerInfo -> containerInfo.getState() !=
+                  HddsProtos.LifeCycleState.DELETED)
+              .map(containerInfo -> containerInfo.getContainerID())
+              .collect(Collectors.toList());
+      DataFilter dataFilter = DataFilter.fromValue(missingIn.toUpperCase());
+
+      switch (dataFilter) {
+
+      case SCM:
+        List<Map.Entry<Long, ContainerMetadata>> notSCMContainers =
+            omContainers.entrySet().stream()
+                .filter(
+                    containerMetadataEntry -> !scmNonDeletedContainers.contains(
+                        containerMetadataEntry.getKey()))
+                .collect(Collectors.toList());
+
+        if (prevKey > 0) {
+          int index = 0;
+          while (index < notSCMContainers.size() &&
+              notSCMContainers.get(index).getKey() <= prevKey) {
```

Comment:
do the notSCMContainers is always ordered? if not, then should skip till it is same if exist. Otherwise need ensure its ordered.

Reply:
Thanks for the comment @sumitagrawl 

The `notSCMContainers` list represents the containers that are missing in the Storage Container Manager (SCM) but exist in the Ozone Manager (OM).
The `getContainerMisMatchInsights` endpoint retrieves information about container discrepancies in the Recon system. The `notSCMContainers` list is generated by iterating over the omContainers, which are fetched using the `reconContainerMetadataManager.getContainers()` API. The `getContainers()` API returns containers in ascending order of container IDs, ensuring that the notSCMContainers list will also be ordered in the same manner.